### PR TITLE
Implement row base exchange in Presto on Spark

### PR DESCRIPTION
**Pull request by arhimondr**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by wenleix** [comment]
Battle tested with 10K mapper & reducer. The query would request ~30TB memory:

Screen Shot 2020-02-17 at 9 46 36 PM

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 1]
```diff
@@ -115,7 +115,11 @@ public static void assertBlockEquals(Type type, Block actual, Block expected)
 
     public static Block createAllNullsBlock(Type type, int positionCount)
     {
-        return new RunLengthEncodedBlock(type.createBlockBuilder(null, 1).appendNull().build(), positionCount);
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 1);
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: unrelated change? -- since the original RLE block is also a valid representation for all nulls block?

File: presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java [hunk 2]
```diff
@@ -17,11 +17,19 @@
...
 
 public interface OutputFactory
 {
-    OperatorFactory createOutputOperator(int operatorId, PlanNodeId planNodeId, List<Type> types, Function<Page, Page> pagePreprocessor, PagesSerdeFactory serdeFactory);
+    OperatorFactory createOutputOperator(
+            int operatorId,
+            PlanNodeId planNodeId,
+            List<Type> types,
+            Function<Page, Page> pagePreprocessor,
+            Optional<OutputPartitioning> outputPartitioning,
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java]: So this `OutputPartitioning` is only required by `PartitionedOutputOperatorFactory` right? (and later `PrestoSparkOutputFactory`)? Maybe check argument in other `OutputFactory` that this shouldn't present

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    public PrestoSparkRow get()
+            throws InterruptedException
+    {
+        PrestoSparkRow row;
+        synchronized (monitor) {
+            while (!finished && buffer.isEmpty()) {
+                monitor.wait();
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: So `PrestoSparkRowBuffer#get` is a blocked call. Is there any reason for that -- I am asking this since the general convention in Presto seems to be using `ListenableFuture<PrestoSparkRow>` for such cases?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+            for (int channel = 0; channel < channelCount; channel++) {
+                Block block = page.getBlock(channel);
+                block.writePositionTo(position, output);
+            }
+
+            boolean shouldReplicate = (replicateNullsAndAny && !hasAnyRowBeenReplicated) ||
+                    nullChannel.isPresent() && page.getBlock(nullChannel.getAsInt()).isNull(position);
+            if (shouldReplicate) {
+                for (int i = 0; i < partitionFunction.getPartitionCount(); i++) {
+                    rowBuffer.enqueue(new PrestoSparkRow(i, output.size(), output.getUnderlyingSlice().byteArray()));
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: I guess you intend to avoid create a new `DynamicSliceOutput` each time so you try to get the underlying byte array directly 😃 . I understand it cannot be done for now since you cannot reset `DynamicSliceOutput` 

That being said, we might want to consider add something like `reset` into `SliceOutput`?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    @Override
+    public void finish()
+    {
+        finished = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished && isBlocked().isDone();
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: is this enough -- do we want to make sure `rowBuffer` is empty?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java [hunk 6]
```diff
@@ -74,18 +81,29 @@ public Page getOutput()
             return null;
         }
 
-        SerializedPage serializedPage = null;
+        PageBuilder pageBuilder = new PageBuilder(types);
         synchronized (iterator) {
-            if (iterator.hasNext()) {
-                serializedPage = iterator.next();
+            while (iterator.hasNext() && !pageBuilder.isFull()) {
+                PrestoSparkRow row = iterator.next();
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java]: As a future work, I guess you want to use Spark's `MutableRow`? :) . cc @sameeragarwal

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java [hunk 7]
```diff
@@ -74,18 +81,29 @@ public Page getOutput()
...
+                SliceInput sliceInput = new BasicSliceInput(wrappedBuffer(row.getBytes(), 0, row.getLength()));
+                pageBuilder.declarePosition();
+                for (int channel = 0; channel < types.size(); channel++) {
+                    BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(channel);
+                    blockBuilder.readPositionFrom(sliceInput);
+                }
+                sliceInput.close();
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java]: looks like this did nothing for `BasicSliceInput` ?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java [hunk 8]
```diff
@@ -123,17 +141,17 @@ public void noMoreSplits()
...
         private boolean closed;
 
-        public SparkRemoteSourceOperatorFactory(int operatorId, PlanNodeId planNodeId, Iterator<SerializedPage> iterator, PagesSerde serde)
+        public SparkRemoteSourceOperatorFactory(int operatorId, PlanNodeId planNodeId, Iterator<PrestoSparkRow> iterator, List<Type> types)
         {
             this.operatorId = operatorId;
             this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
             this.iterator = requireNonNull(iterator, "iterator is null");
-            this.serde = requireNonNull(serde, "serde is null");
+            this.types = ImmutableList.copyOf(requireNonNull(types, "serde is null"));
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java]: nit: `types is null`

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java [hunk 9]
```diff
@@ -270,11 +258,20 @@ private PrestoSparkQueryExecution(
...
-                    .map(this::deserializePage)
-                    .flatMap(page -> getPageValues(connectorSession, page, rootFragment.getTypes()).stream())
-                    .collect(toList());
+            List<Type> types = rootFragment.getTypes();
+            ImmutableList.Builder<List<Object>> result = ImmutableList.builder();
+            for (Tuple2<Integer, PrestoSparkRow> tuple : resultRdd) {
+                PrestoSparkRow row = tuple._2;
+                SliceInput sliceInput = new BasicSliceInput(Slices.wrappedBuffer(row.getBytes(), 0, row.getLength()));
+                ImmutableList.Builder<Object> columns = ImmutableList.builder();
+                for (Type type : types) {
```
Comment (wenleix) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java]: I am wondering if we should have a utility function to convert a list of `PrestoSparkRow` into a list of Blocks? -- we can leave this as TODO .

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java [hunk 10]
```diff
@@ -17,18 +17,44 @@
 
 import static java.util.Objects.requireNonNull;
 
-public class SerializedPrestoSparkPage
+public class PrestoSparkRow
         implements Serializable
 {
+    private static final int INSTANCE_SIZE = Long.BYTES * 2 /* headers */
```
Comment (wenleix) [inline presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: Why not using `ClassLayout.parseClass` as we did for other Presto class? -- I am asking this is just adding these bytes together doesn't necessarily sum up to the instance size ,  e.g. there are Java object padding size, etc.

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java [hunk 11]
```diff
@@ -17,18 +17,44 @@
...
+    }
+
     public byte[] getBytes()
     {
         return bytes;
     }
+
+    public long getRetainedSize()
+    {
+        return INSTANCE_SIZE + bytes.length;
```
Comment (wenleix) [inline presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: Similarly, we usually use `SizeOf.sizeOf(bytes)`

**Comment by wenleix** [comment]
Travis failure is relevant.

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java (hunk 1 above)
Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: It just feels a little contrintuitive. I would expect the `createAllNullsBlock` to create a native block for a given type when all positions are nulls. If the intent is to create a RLE block with a null value - there should be a separate method that does that.

Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: Let me extract it into a separate commit

**Comment by arhimondr** [comment]
> Speaking of the method name, what about serializePosition and deserializePosition? cc @tdcmeehan

I actually like it better

File: presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java (hunk 2 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java]: Added a check in the `TaskOutputFactory`. Didn't add for others, as the others are only used in tests.

**Comment by arhimondr** [comment]
> For the commit message, maybe also mentioned refactor partitioned output related information into OutputPartitioning ?

Added more detailed description of what am I trying to achieve by doing the refactoring.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final PrestoSparkRowBuffer rowBuffer;
+        private final Function<Page, Page> pagePreprocessor;
+        private final PartitionFunction partitionFunction;
+        private final List<Integer> partitionChannels;
+        private final List<Optional<Block>> partitionConstants;
+        private final boolean replicateNullsAndAny;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: > what does replicateNullsAndAny mean?

It is needed for the SemiJoin. It tels the partitioner to copy null rows to every partition and also make sure that if the input is not empty - a row is replicated to every partition (so the partitions are not empty). It is needed to preserve correct `null` vs `false` semantics of the SemiJoin.

I think the `replicatesAnyRow` is missleading. Whoever was extending this behaviour to the `replicateNullsAndAny` probably just forgot to do the rename in the PartitioningOperator.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java (hunk 3 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: The Spark iterator is blocking anyway. It is mostly for implementation simplicity. Otherwise we would always wait on the `Future#get()` unconditionally in the SparkIterator.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java (hunk 4 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: This should be done as part of the buffer recycling effort. As SliceOutput has to acquire a free buffer from a pool before you can reset it.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java (hunk 6 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java]: Actually I'm not sure if we want to integrate with Spark's types at this point.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java (hunk 7 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRemoteSourceOperator.java]: Just in case =)

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java (hunk 10 above)
Comment (arhimondr) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: The Jol library is not in the Spark's classpath, and I don't want to add one to avoid potential clashes.

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java (hunk 11 above)
Comment (arhimondr) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: SizeOf is in `Slice` library. I don't want to add it to avoid potential clashes.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        @Override
+        public OperatorFactory createOutputOperator(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<Type> types,
+                Function<Page, Page> pagePreprocessor,
+                Optional<OutputPartitioning> outputPartitioning,
+                PagesSerdeFactory serdeFactory)
+        {
+            OutputPartitioning partitioning = outputPartitioning.orElse(SINGLE_PARTITION);
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: > For example, in

Github must've seved not the final version. Could you pleas elaborate?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java (hunk 5 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: It looks like the `TaskOutputFactory` is not waiting for the buffer to be drained. I followed the same approach here.

**Comment by arhimondr** [comment]
@wenleix Comments addressed

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java [hunk 14]
```diff
@@ -270,11 +258,20 @@ private PrestoSparkQueryExecution(
...
+                    BlockBuilder blockBuilder = type.createBlockBuilder(null, 1);
+                    blockBuilder.deserializePosition(sliceInput);
```
Comment (tdcmeehan) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java]: Have you considered adding something like an overload to `getObjectValue` that takes in a `sliceInput`, to avoid having to create the block builder per cell?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+public class PrestoSparkRowBuffer
+{
+    private final OutputBufferMemoryManager memoryManager;
+
+    private final Object monitor = new Object();
+    @GuardedBy("monitor")
+    private final Queue<PrestoSparkRow> buffer = new ArrayDeque<>();
+    @GuardedBy("monitor")
+    private volatile boolean finished;
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: Yes. Since the variable is guarded by lock the volatile modifier is not needed, as the synchronized block guarantees the ordering. However it is generally a good practice to keep the `volatile` modifier as it doesn't introduce any performance regressions while allowing to use the variable directly without synchronization where compare-and-set semantics are not needed.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkRowBuffer
+{
+    private final OutputBufferMemoryManager memoryManager;
+
+    private final Object monitor = new Object();
+    @GuardedBy("monitor")
+    private final Queue<PrestoSparkRow> buffer = new ArrayDeque<>();
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: `ArrayBlockingDeque` blocks based on the number of elements. We want to block based on the memory utilization instead.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        memoryManager.updateMemoryUsage(row.getRetainedSize());
+    }
+
+    public void setNoMoreRows()
+    {
+        memoryManager.setNoBlockOnFull();
+        synchronized (monitor) {
+            finished = true;
+            monitor.notifyAll();
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: No. Buffered rows has to be consumed by the consumer.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java (hunk 14 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java]: This is only used to produce the final output to the client, which is not expected to be large.

**Comment by tdcmeehan** [comment]
> "Implement readPosition and writePosition in Block/BlockBuilder"
> 
> LGTM as an initial implementation. I am sure there are a lot of optimization space (especially `invokeinterface` is quite expensive).
> 
> Speaking of the method name, what about `serializePosition` and `deserializePosition`? cc @tdcmeehan

I actually preferred the earlier names.  Adding something serialization related to Block and BlockBuilder isn't very flexible, because it's taking a hard position also on the serialization format without a way to customize it.  The name even made me consider if it was worthwhile to put these methods onto BlockEncodingSerde.  But I don't feel super strongly about it.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: I think for boolean/byte/integer, JMM grantees it's an atomic update: https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.7 . 

@arhimondr : Are you more worried about the "Happens Before" memory model ? (See references in https://github.com/prestodb/presto/pull/8772#issuecomment-323800658)  . In that case adding `volatile` guarantees sequentially consistent  memory model :)

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: Discussed offline, a linked blocking queue would work but we'd suffer pointer indirection costs.  There's high performance unbounded array blocking queues out there but not worth the additional dependency.

File: presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java (hunk 2 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/OutputFactory.java]: Might still be useful to explicitly add the tests (it's more for convenient for readers of the code :)  )

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java]: Since this variable is never accessed outside the lock it doesn't have to be volatile as synchronization enforces the memory ordering. It can be safely removed. I added it just in case somebody accidental uses it outside the lock (e.g.: simple atomic read). However now I'm not sure how much of sense does it make. Let me simply remove it to avoid confusion.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java]: @arhimondr : @arhimondr : For example, in `TaskOutputOperatorFactory` ,  the output partitioning can be any of the `FIXED_BROADCAST_DISTRIBUTION`, `FIXED_ARBITRARY_DISTRIBUTION`, `SCALED_WRITER_DISTRIBUTION`, `SINGLE_DISTRIBUTION`, `COORDINATOR_DISTRIBUTION`: https://github.com/prestodb/presto/blob/2fb54ed0dc380791876810237c116f587efc2470/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L381-L395

Update: with the check argument that output partitioning must be absent, it might be OK... just note the absence of output partitioning seems to have different semantic for `TaskOutputOperatorFactory` vs. `PrestoSparkOutputFactory`

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java (hunk 11 above)
Comment (wenleix) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: In this case, should we just copy `sizeOfByteArray` implementation? i.e. something like 

```java
    public long getRetainedSize()
    {
        return INSTANCE_SIZE + sizeOfByteArray(bytes.length);
    }

    // Copied from SizeOf#sizeOfByteArray in Slice library
    private static long sizeOfByteArray(int length)
    {
        return ARRAY_BYTE_BASE_OFFSET + (((long) ARRAY_BYTE_INDEX_SCALE) * length);
    }
```

**Comment by arhimondr** [comment]
> The name even made me consider if it was worthwhile to put these methods onto BlockEncodingSerde

I was thinking about something similar. However It feels like it won't play well with the way `BlockEncodingManager` is used. It is created as a singleton, and it assumes that the block type is going to be encoded in the input. 

When serializing row-by-row the schema is not encoded in the serialized representation, thus the BlockBuilders must be pre-created based on the input types before decoding.

I was also thinking about creating a separate abstraction, similar to `BlockEncodingSerde`, but then it will have to be integrated with connectors / plugin manager / etc, which seems to be an overkill.


@wenleix @tdcmeehan Thoughts?

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java [hunk 18]
```diff
@@ -17,18 +17,44 @@
...
 
-public class SerializedPrestoSparkPage
+public class PrestoSparkRow
         implements Serializable
 {
+    private static final int INSTANCE_SIZE = Long.BYTES * 2 /* headers */
+            + Integer.BYTES /* partition */
+            + Integer.BYTES /* length */
+            + Long.BYTES /* bytes pointer */
+            + Long.BYTES * 2 /* bytes headers */
```
Comment (wenleix) [inline presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: bytes headers and bytes length should probably go into `sizeOfByteArray`?

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java (hunk 10 above)
Comment (wenleix) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: @arhimondr : Sounds reasonable, just add a comment explain why this is the case. 

Also, looks like JOL and your calculation returns different results? (You seems to over-estimate): https://gist.github.com/wenleix/faae51ac5fe3c0f688d8520c7c0d3105

```
Instance Size from JOL: 24
Instance Size from Andrii: 32
```

**Comment by wenleix** [comment]
> When serializing row-by-row the schema is not encoded in the serialized representation, thus the BlockBuilders must be pre-created based on the input types before decoding.

> I was also thinking about creating a separate abstraction, similar to `BlockEncodingSerde`, but then it will have to be integrated with connectors / plugin manager / etc, which seems to be an overkill.

Makes sense. Let's mark these two methods with `@Beta` for now. Looks to me `BlockEncodingSerde` is designed for serialize the whole block, and doesn't seem to work well with the row-by-row serialization (unless creates a separate abstraction) .

Also different from Block granularity serialization ,  the most common operation in row-by-row serialization is to "drill down" the delegate blocks, which seems to more natural to `Block` interface (think about `writePositionTo`, etc.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java (hunk 9 above)
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java]: Currently this is a single place where it is done. Let's keep it here for now.

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java (hunk 18 above)
Comment (arhimondr) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkRow.java]: sizeOfByteArray is in the Slice library. It is not available here =\

**Comment by arhimondr** [comment]
@wenleix Any opinion on the naming?

**Comment by wenleix** [comment]
Required by https://github.com/prestodb/presto/issues/13856
